### DISCUSSION TITLE
Dynamically adding Disks/Filesystems

### DISCUSSION THREAD

**Discussion by dsdole** [siderolabs/talos#9894] [Q&A] [answered]
# Idea
I am working on some management scripts and I was wondering about the following scenario:
- A single talos node (known as **TALOS** hosted as a Proxmox VM. The talos node has a single 128GiB virtual disk for the system install.
- There is no persistent volume provisioner configured on the k8s cluster.
- There is a "storage management" vm (known as **MGMT**) includes tools for initializing block device, creating filesystems,..., etc.
- There is my "control software" known as **DAEMON**. **DAEMON** has full control over the Proxmox cluster, **MGMT** and **TALOS**

I want to manage Kubernetes Persistent Volumes as Proxmox/Qemu virtual disks. I have thought of the following workflow:

## Provisioning a Volume
1. **DAEMON**: receives a request to deploy an application which needs a PV(Persistent Volume) with specs: {size, fs, io_limits}
2. **DAEMON**: uses Proxmox API to create a virtual disk and attaches it to **MGMT**.
3. **MGMT**: Creates the filesystem and any required directories.
4. **DAEMON**: Detaches virtual disk from **MGMT** and attaches it to **TALOS**.
5. **DAEMON**: Generates Talos machine config patch or api commands and makes **TALOS** mount the virtual disk's filesystem at a specified location "/srv/pv/volume"
6. **DAEMON**: Deploys a K8s deployment to **TALOS** which uses a host path bind mount.

# Why
My goal here is to treat PVs just like block devices on the public clouds: I want to be able to shift them around, and back them up via Proxmox Backup Server.

# Questions
- Any general thoughts on this?
- Will I run into weird behaviors when repeatedly attaching and detaching disks to TalosLinux.
- Will I run into weird behaviors with mounting the Filesystems?
- Once a disk is attached, If I generate a config patch to mount the filesystem, would the talos node need to reboot to apply the changes?

**Comment by smira** [comment, marked as answer]
This sounds like e.g. https://github.com/sergelogvinov/proxmox-csi-plugin 

In general, google for 'Proxmox CSI'.
